test/integration/cluster/cluster_test.go

```diff
@@ -0,0 +1,105 @@
+package deployment
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	informers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"os"
+	clusterv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	clientset "sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+	client "sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset/typed/cluster/v1alpha1"
+	"testing"
+)
+
+func TestCluster(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Cluster-Controller")
+}
+
+var clusterInstance = &clusterv1alpha1.Cluster{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      "foo",
+		Namespace: "clusterapi-test",
+	},
+	Spec: clusterv1alpha1.ClusterSpec{
+		ClusterNetwork: clusterv1alpha1.ClusterNetworkingConfig{
+			ServiceDomain: "mydomain.com",
+			Services: clusterv1alpha1.NetworkRanges{
+				CIDRBlocks: []string{"10.96.0.0/12"},
+			},
+			Pods: clusterv1alpha1.NetworkRanges{
+				CIDRBlocks: []string{"192.168.0.0/16"},
+			},
+		},
+	},
+}
+
+var testNamespace = "clusterapi-test"
+
+var _ = Describe("Cluster-Controller", func() {
+	var clusterapi client.ClusterInterface
+	var client *kubernetes.Clientset
+	var stopper chan struct{}
+	var informer cache.SharedIndexInformer
+
+	BeforeEach(func() {
+
+		// Load configuration
+		kubeconfig := os.Getenv("KUBECONFIG")
+		config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+		Expect(err).ShouldNot(HaveOccurred())
+
+		// Create kubernetes client
+		client, err = kubernetes.NewForConfig(config)
+		Expect(err).ShouldNot(HaveOccurred())
+
+		// Create namespace for test
+		ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: testNamespace}}
+		_, err = client.Core().Namespaces().Create(ns)
+		Expect(err).ShouldNot(HaveOccurred())
+
+		// Create  informer for events in the namespace
+		factory := informers.NewSharedInformerFactoryWithOptions(client, 0, informers.WithNamespace(testNamespace))
+		informer = factory.Core().V1().Events().Informer()
+		stopper = make(chan struct{})
+
+		// Create clusterapi client
+		cs, err := clientset.NewForConfig(config)
+		Expect(err).ShouldNot(HaveOccurred())
+		clusterapi = cs.ClusterV1alpha1().Clusters(testNamespace)
+	})
+
+	AfterEach(func() {
+		close(stopper)
+		client.Core().Namespaces().Delete(testNamespace, &metav1.DeleteOptions{})
+	})
+
+	Describe("Create Cluster", func() {
+		It("Should trigger an event", func() {
+			// Register handler for cluster events
+			events := make(chan *corev1.Event, 0)
+			informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					e := obj.(*corev1.Event)
```

Comment:
Do you want to check if this is an Event and fail if it's not?

Reply:
The intention of this skeleton is to show the general form of the tests: inform of events related to the cluster and check those are the expected events. The actual verification depends of each specific test. Also, it requires the events carry the required information, something that has still to be reviewed in the test provider.

Reply:
If someone accidentally coded `informer = factory.Core().V1().Pods().Informer()` (pods instead of events), this type assertion would fail. I would say it's safer to validate the type of `obj`.

Reply:
I still think it's worth validating the received type. Something like
```go
Expect(obj).Should(BeAssignableToTypeOf(&corev1.Event{}))
```
... if that's the correct gomega matcher syntax.

Reply:
@ncdc The informer is created for events: `informer = factory.Core().V1().Events().Informer()`.

I expect only Events to be received. Wouldn't checking type be redundant?

Reply:
I agree with you that it is unlikely you would get an incorrect type, but it is technically possible due to programmer error (using the wrong informer). The informer framework does not have static typing guarantees, which is why I suggested adding the type check. It may sound redundant, but I'd rather have an `Expect` failure than a panic in the event of an accidental mismatch between the informer and the expected type. However, I'm not going to hold this PR up because of this 😄